### ISSUE TITLE
[BUG] Can't use component testing in existing project

### ISSUE THREAD

**Issue by piotrpalek** [microsoft/playwright#17804]
**Context:**
- Playwright Version: 1.26.1
- Operating System: Mac
- Node.js version: 16.13.0
- Browser: All

CodeSandbox link showing bug (running below repo): https://codesandbox.io/p/github/piotrpalek/template-hcp/playwright
Github link to example repo: https://github.com/piotrpalek/template-hcp/tree/playwright

**Describe the bug**

Hey, I'm trying to run playwright, on an existing webpack project, but I'm getting an issue when trying to import modules without adding a file extension. It's saying "Error: Cannot find module" when trying to run it. I've succeeded in making an example repo showing the exact same issue.

It would be pretty cumbersome (and against our current styleguide) to change all imports to have file extensions. Is there a workaround I could use here?

**Comment by yury-s** [comment]
> Github link to example repo: https://github.com/piotrpalek/template-hcp/tree/playwright

I cloned this repo and run the commands you mentioned but I got the error:

```
$ npm run dev
npm ERR! Missing script: "dev"
```

**Comment by piotrpalek** [comment]
@yury-s sorry the instructions were from a template, I've updated the repo with the correct one, fwiw you just need to run `yarn run test-ct` (or `yarn run start` in that repo which does the same)

**Comment by yury-s** [comment]
> @yury-s sorry my bad, the instructions were from a template, I've updated the repo with the correct one :) you need to run `yarn run test-ct` (or `yarn run start` in that repo which does the same)

Still not working:
```
error Command "test-ct" not found.
```

**Comment by yury-s** [comment]
> edit: I've pushed a proof of concept PR that seems to fix my issue over here: #17839

We need to be able to reproduce the problem before accepting a fix, maybe you can add a test to our suite that demonstrates the bug if it's easier?

**Comment by yury-s** [comment]
> @yury-s that's weird I just tried the repo from scratch and `yarn run test-ct` works. Did you clone it from scratch? I've changed the link to point to a fresh repo to make sure.

Oh, you updated the link for CodeSandbox, I was using the second one. I can reproduce it with https://github.com/piotrpalek/playwright-issue

**Comment by pavelfeldman** [comment]
@piotrpalek I can see you are using the type: module. This mode requires that all the imported paths have extensions, by the specification. Furthermore, for imported TS paths you should be using JS extensions. Any reason you want to use type: module?

See https://www.typescriptlang.org/docs/handbook/esm-node.html

**Comment by piotrpalek** [comment]
> Any reason you want to use type: module?

@pavelfeldman that's a good question, I think I've hit some issue with importing, searched around issues and came to the conclusion I need to use `type:module` for playwright to work.

I think what triggered this was that whole chain is the following error when I add `lodash-es` to the test, which in my actual project is imported somewhere down the line (I've removed `type:module` here):
![image](https://user-images.githubusercontent.com/1474323/194242673-dd1b40df-7ed8-4fd3-857f-1e0170940534.png)

I've uploaded this to a new branch called [commonjs](https://github.com/piotrpalek/playwright-issue/tree/commonjs) on that same repo.

**Comment by pavelfeldman** [comment]
> When I add lodash-es to the test this error pops up:

That's because you picked an ESM version of that popular module.

**Comment by pavelfeldman** [comment]
> that itself imports another .js file via an alias

JavaScript has no notion of an alias, runtime would be resolving these imports unless there is a bundler with the configured aliases, which I suspect is happening in your case. You should pass these aliases as a part of the Vite config using the ctViteConfig property. Check out the alias property in https://playwright.dev/docs/next/test-components#q-how-can-i-use-vite-plugins.

**Comment by pavelfeldman** [comment]
> if there's a tsconfig, shouldn't all files always go through the transform

tsconfig only applies to Typescript.

**Comment by piotrpalek** [comment]
> tsconfig only applies to Typescript

it gets tricky/muddy though when we're importing ts files from js files, and vice versa

**Comment by pavelfeldman** [comment]
> yep :) just wanted to show an example where the project works under webpack, but there's difficulty making it work with playwright, what would be the play here?

Webpack considers your source to be ESM, so it processes ESM imports properly. As if you had type: module, in which case Playwright, as a part of Node.js runtime would process them properly.

> it gets tricky/muddy though when we're importing ts files from js files, and vice versa

TS get transpiled to JS, so when you say "importing ts files from js files and vice versa", what really happens is you have js files importing js files and vice versa at runtime.

> Is the config in the example repo wrong then? I'm importing import nestedNonTsUtil from 'spec-utils/nestedNonTsUtil', the path of the file is spec/javascripts/utils/nestedNonTsUtil.js

Ah, we were talking about different things here. When Playwright tests your components, it builds them first using Vite. If the source code for those components uses aliases, you would specify them in ctViteConfig. That way Vite knows how to build your components. The code that Vite built will run in your browser, it has nothing to do with your tests so far.

Now the tests themselves run in Node, they have filesystem access, etc. Node does not have a notion of aliases, so it'll be looking for `spec-utils` node module when you put absolute require like yours in code. Typescript has a notion of aliases, defined in tsconfig, so if your nonTsUtil.js was TS, things would work. Which I think you figured out yourself given the file name you used.

tl/dr: Node has no concept of aliases, only transpilers and bundlers do (Babel, rollup, etc). When you work on components you already have vite / rollup / webpack /babel config with those aliases defined, include them in your ctViteConfig. The rest is pure Node with TS transpilation that would work by the spec.

**Comment by piotrpalek** [comment]
I think I more or less get all of that now, thanks for the clarification!

The issue is that it's very impractical at this point for us to try to separate code between "test code that runs in nodeland and can't use aliases" and "code that's related to components and is transpiled hence knows aliases", as they tent to import from each other (even if it's some deeply nested relationship).

So in effect I can't use playwright the way I could Jest or Cypress. In jest I can specify the aliases, and Cypress allows me to run the webpack devserver for my code. 

What would the solution be for Playwright? I'd really rather use Playwright, but it seems I can't get it to work in my setup, or don't know how without hacking the loader and getting it to respect the aliases even in nodeland

**Comment by pavelfeldman** [comment]
> I guess what's mostly confusing for me is that part of the code get's the Vite treatment, and part of it does not, hence the incompatibilities I'm facing

That's exactly right. In order to be efficient (load web pages quickly, allow parallel test execution, enable filesystem access and other superpowers in tests), we've separated the test runner from the web page. In cases of Jest and Cypress both run in the same environment (Node and Browser respectively).

It seems that the problem is that (due to the utilities reuse), you are importing the code that can't run in Node, it was only intended to be bundled and run in the browser. We control module resolution, so we can introduce aliases even for Node.js environment. But a) it won't be pure Node anymore and b) the problem is broader than that. For example, imagine that your nonUsUtil.js file uses some `window.` APIs. It would be totally fine to include them in your component / web app code, but you can't be using them outside of the browser. So even if we fix the aliases problem for you, you should be mindful of only reusing isomorphic code in those utilities.

So yes, Playwright makes you think about the tests as 'My components run in the browser while my tests run in Node', and yes it is more involved than running your test runner inside the web page (Cypress) or running your web component in the simulated environment in Node (Jest). But at the same time it is faster, it is more powerful (exposes Node APIs to the tests) and more resilient (bad pages can't destroy your test run). You should pick the tool that works better for you. It looks like the cost of adopting Playwright is along the lines with replacing aliases with relative imports and/or renaming those .js files into .ts. I realize it is not trivial, so you have a call to make.

**Comment by piotrpalek** [comment]
Ok this is making sense to me now, I just had a totally different concept of how Playwright works. It's also good to know how it's supposed to work, as I was wondering if what I'm encountering is a bug or a deliberate design choice :)

> We control module resolution, so we can introduce aliases even for Node.js environment (...) So even if we fix the aliases problem for you, you should be mindful of only reusing isomorphic code in those utilities

Yes that's true, but I feel that's an easier thing to solve on the project side. I find that separating how aliases work on .ts vs .js fiels is confusing, even if we only consider the node.js environment. In my experience if you're using typescript it's very likely the aliases are used for both .js and .ts files (as probably all of them are pushed through the tsc compiler or some bundler at that point)

> a) it won't be pure Node anymore

I'd argue that since you're taking into account a `tsconfig` and support typescript it's not pure node already. Now for my case (and I'd think for many others) it would be better if at that point we let all files through typescript. But I understand that there may be cases where that's not desirable. I'm not sure what the solution here is, but I think both use cases are valid and should be possible. 

Maybe (optionally) using Vite on the Nodejs/Test side and letting us configure both separately would be some sort of solution here?

I think for me this is solved, I now know how why I'm having issues and that this isn't some bug in playwright, but just how it works now, and I need to change a few things in the project in order for it to work. Thanks again y'all for your time, feel free to close this issue if it makes sense to you :)

**Comment by pavelfeldman** [comment]
I think we are on the same page now.

In terms of actionability on this bug, configuring aliases is somewhat a mess in the Node land: different bundlers will use different config files for that, they won't respect the tsconfig's ones. In the end every framework ends up introducing its own config. Here is an [example](https://divotion.com/blog/how-to-configure-import-aliases-in-vite-typescript-and-jest). Our options would be to:
- A: introduce aliases in our config
- B: respect aliases from tsconfig for JS files
- C: do both: introduce aliases and fall back to tsconfig if there is any

I don't think we should do B since it is puling JS people into the TS land for just that feature. So it would need to be something like C.
